### ISSUE TITLE
Unable to request additional data (email, friends etc.) with API v2.10

### ISSUE THREAD

**Issue by mitianre** [jaredhanson/passport-facebook#214]
I guess this is somehow related to [#198](https://github.com/jaredhanson/passport-facebook/issues/198). I've created a new app and the minimum API version is 2.10. The only fields that I'm able to get are the id, name and profile picture.

I've tried all the solutions to get other data, profileFields with 'email' or 'emails' and with scope setted for 'email' (regarding the email), also tried to set authorization, profile and token URL to the new API with no success.

And it is not a problem on my side, I've also tested my profile with Graph API Explorer and I'm able to get any data I want with the right permissions and scope.

**Comment by riston** [comment]
I had a similar issue, with the new users which were unable to get the email field. Make sure when you get the permission request dialog the email is added there, otherwise you need to revoke the app permission and do it again. 

Here are my settings:
```      
scope:
- 'email'
- 'public_profile'
profile_fields:
- 'id'
- 'email'
- 'first_name'
- 'last_name'
- 'middle_name'
- 'gender'
- 'link'
```

**Comment by mitianre** [comment]
I've revoked the permissions for the app and when I do a new request, the permissions dialog only asks for public profile, nothing related to email or other data I request.

I've used the same settings as you, except middle_name, gender and link, and with some additions (list of friends that also use my app).

**Comment by benjaminadk** [comment]
I noticed i could not get Email even though i included:
`profileFields: ['id', 'displayName', 'picture', 'email']`
in my **new FacebookStrategy** config object.
However, I also had to add the following to my facebook authentication route:
```
server.get('/auth/facebook',
  passport.authenticate('facebook', { scope: ['email']})); 
```
NOTE: i am calling my `express()` instance **server** instead of **app**.  it makes no difference, I could call it **app** and it would work the same way...just for clarity if someone reads this.
Once I did this I was prompted with a new message when activating my **Signup with Facebook** route.
I was being asked if my Application could access **email** from Facebook.  
Since **public_profile** is always included I still got all the original information, plus now, I was able to access the email address within the user's Facebook Profile.
I am not saying anything different from @riston, just providing what some of my code looked like, and confirming I had the same issue initially. Hopefully, @mitianre has things cleared up by now, if not, hope this helps.

**Comment by mitianre** [comment]
My settings are something like this (+ clientId, secret & callbackUrl):

`{
            authorizationURL: 'https://www.facebook.com/v2.10/dialog/oauth',
            tokenURL: 'https://graph.facebook.com/v2.10/oauth/access_token',
            profileURL: 'https://graph.facebook.com/v2.10/me',
            profileFields: ['id', 'email', 'first_name', 'last_name', 'friends', 'location', 'picture.width(200).height(200)']    
 }`

And my auth route:

`authRoutes.get('/login/facebook', passport.authenticate('facebook', {session: false, scope: ['email', 'user_friends', 'user_location']}));`

And it's not working.

**Comment by riston** [comment]
@mitianre when the FB permission request dialog window opens, make sure the scope that you set exists also there. If you already have accepted the permissions with only `public_profile`, you have to revoke the application permissions (https://www.facebook.com/settings?tab=applications), then you can accept it again.

**Comment by mitianre** [comment]
That's the thing, the dialog asks just for public_profile. I went through the process of revoking and accepting it again, and it has the same behaviour.

**Comment by riston** [comment]
When you are redirected to FB the URL includes the correct scope? As you can see the scope includes `email and public_profile`: `https://www.facebook.com/dialog/oauth?response_type=code&redirect_uri=https%3A%2F%2Fcallback_changed%2Foauth%2Fcallback%2Ffacebook&scope=email%2Cpublic_profile&state=cj6x9cin5000204xbape1y13l&client_id=xxxxxxxxxxxxxxxxxx`

With Chrome Dev tools enable the Preserve log option in Network tab to see the URL.

**Comment by mitianre** [comment]
Apparently there is no scope in the URL.

**Comment by kennasoft** [comment]
I'm also experiencing the same problem with a new app using facebook's API version 2.10. It's likely a bug that the maintainers will need to look into. Maybe something has changed in facebook's API implementation that they need to keep up pace with.

**Comment by kennasoft** [comment]
@mitianre I eventually solved it by adding `profileFields` to the options object. I looked through the code for facebook strategy, but couldn't find any reference to the `scope` object (maybe it's used at another layer). I think it would be nice to update the docs to let people know that scope doesn't work anymore.

Here's a sample of what I used

```javascript
let fbStrategy = new FacebookStrategy({
                clientID: '[my_client_id]',
                clientSecret: process.env.fbsecret,
                callbackURL: '[my callback URL]',
                profileURL: 'https://graph.facebook.com/v2.10/me',
                authorizationURL: 'https://www.facebook.com/v2.10/dialog/oauth',
                tokenURL: 'https://graph.facebook.com/v2.10/oauth/access_token',
                profileFields: ['email','first_name','last_name','gender','link']
            },
    function(token, refreshToken, profile, done){
         //find local user with matching email...
         ...etc
    });
```

This returns all the fields specified in the `profile` object.

**Comment by mitianre** [comment]
I already had profileFields inside my options object. It's not working even if I use scope or not.

**Comment by pmichael16** [comment]
I was able to make mine work with these settings. Thanks to @kennasoft! I just added a few more perms same as yours @mitianre. The login dialog displays this "_Some App_ will receive:
your public profile, friend list, current city and email address".

PassportJS Config:
```
passport.use(
    new FacebookStrategy(
      {
        clientID: keys.fbClientID,
        clientSecret: keys.fbClientSecret,
        callbackURL: '/auth/facebook/callback',
        profileURL: 'https://graph.facebook.com/v2.10/me',
        authorizationURL: 'https://www.facebook.com/v2.10/dialog/oauth',
        tokenURL: 'https://graph.facebook.com/v2.10/oauth/access_token',
        profileFields: ['id', 'displayName', 'picture.width(200).height(200)', 'first_name', 'middle_name', 'last_name', 'gender', 'link', 'email', 'location', 'friends']
      },
      (accessToken, refreshToken, profile, done) => {
        console.log('accessToken: ', accessToken);
        console.log('refreshToken: ', refreshToken);
        console.log('profile: ', profile);
      }
    )
  );
```
AuthRoute:
```
app.get(
    '/auth/facebook',
    passport.authenticate('facebook', {
      scope: ['user_location', 'email', 'user_friends']
    })
  );
```

The way I see it, the **profileFields** value should correspond to what's indicated in [Facebook's Graph API Reference](https://developers.facebook.com/docs/graph-api/reference) and the **scope** value should correspond to [Facebook's Login Permissions](https://developers.facebook.com/docs/facebook-login/permissions/).

I hope this kind of solution will be updated in the docs soon --most especially how to use Graph API's latest version (v2.10) I only knew how by reading through this discussion. Thanks guys! :)

**Comment by mitianre** [comment]
For me it's not working even if I use the same config. I have no idea what's wrong. I don't know, do you have your app (in facebook developers dashboard) public? Mine is in development mode, but I'm testing it on my account.

**Comment by pmichael16** [comment]
Mine is still in development mode and I also use my personal account to test. Have you tried using the same settings I posted above? Do you see any error in your terminal? Maybe you could find a lead there. 

By the way, make sure to remove your App in your [FB App Settings](https://www.facebook.com/settings?tab=applications) so you can see the login dialog box again to re-authorize your App. 

If you followed what I did, once you go to '/auth/facebook' you should see the login dialog and as you click "Continue" you should see your profile logged in your terminal where you run Node.

You should see something like this:

```
accessToken:  _______
refreshToken:  undefined
profile:  { id: '_______',
  username: undefined,
  displayName: '_______',
  name: 
   { familyName: '_______',
     givenName: '_______',
     middleName: '_______' },
  gender: '_______',
  profileUrl: 'https://www.facebook.com/app_scoped_user_id/_______/',
  emails: [ { value: '_______@_______.com' } ],
  photos: [ { value: 'https://scontent.xx.fbcdn.net/_______' } ],
  provider: 'facebook',
  _raw: '{"id":"_______","name":"_______","picture":{"data":{"height":200,"is_silhouette":false,"url":"https:\\/\\/scontent.xx.fbcdn.net\_______","width":200}},"first_name":"_______ _______","last_name":"_______","gender":"_______","link":"https:\\/\\/www.facebook.com\\/app_scoped_user_id\\/_______\\/","email":"_______.com","location":{"id":"_______","name":"_______"},"friends":{"data":[],"summary":{"total_count":_______}}}',
  _json: 
   { id: '_______',
     name: '_______',
     picture: { data: [Object] },
     first_name: '_______',
     last_name: '_______',
     gender: '_______',
     link: 'https://www.facebook.com/app_scoped_user_id/_______/',
     email: '_______@_______.com',
     location: { id: '_______', name: '_______' },
     friends: { data: [], summary: [Object] } } }
```
The actual output has texts in the id, name, etc. The blanks '_______' there are just for demonstration purposes.

**Comment by mitianre** [comment]
Nah, I have no error, it's just not working.  As previously discussed in the previous comments, there is no scope included in the URL when redirected to facebook and I don't know why.

**Comment by polarity** [comment]
i tried the graph explorer for the first time and i get something like this:
![graph api explorer - facebook for developers](https://user-images.githubusercontent.com/147424/31292012-14fb4292-aad2-11e7-90d8-c8d4ce131b4d.png)

1.) renewing the access token didn't helped (allowing access to email)
2.) updating the email didn't helped

any ideas?

**Comment by polarity** [comment]
I changed my email address for my fb profile and now i can retrieve the email via the api.

![graph api explorer - facebook for developers-1](https://user-images.githubusercontent.com/147424/31292591-5a80deba-aad4-11e7-972b-ce5366976fbd.png)

the old email address was "spam@mydomain.de" and i changed it to my primary address "polarity@mydomain.de". Looks like facebook is flagging some email addresses internally as spam.

**Comment by indrekru** [comment]
@mitianre Did you find a fix?

**Comment by mitianre** [comment]
No, I didn't have the time to investigate further. I'm not sure when and how I'm going to make it work. :))

**Comment by statico** [comment]
**TL;DR:** After much debugging I found the facebook-passport wasn't sending the `scope` parameter to OAuth2 at the right time. I solved this problem by moving the `scope` option from the options in my `/auth/facebook/callback` route to my `/auth/facebook` route, like this:

```javascript
  router.get('/auth/facebook', (ctx) => {
    return passport.authenticate('facebook', {
      scope: ['email', 'public_profile'],
      authType: 'rerequest'
    }, (err, user, info, status) => {
      ...
    })(ctx)
  })

  router.get('/auth/facebook/callback',
    passport.authenticate('facebook', {
      failureRedirect: '/error',
      successRedirect: '/'
    }))
```

**Long story:** I debugged this by printing out the access code [at the top of Strategy.prototype.userProfile](https://github.com/jaredhanson/passport-facebook/blob/82a5d8fb04ec915505a8f46565f511bedaad4bb4/lib/strategy.js#L135), trying it on Facebook's [Graph API Explorer](https://developers.facebook.com/tools/explorer/), and confirming that the email was missing:

![image](https://user-images.githubusercontent.com/137158/33510456-fe06a23c-d6c0-11e7-99da-2921a5d212da.png)

(In the explorer I chose API v2.5 explicitly because that's what [passport-facebook currently uses](https://github.com/jaredhanson/passport-facebook/blob/82a5d8fb04ec915505a8f46565f511bedaad4bb4/lib/strategy.js#L56). It doesn't make a difference though.)

However, if I use the *Get User Access Token* button on that page, request `email` access, and run the same query, I get my email address!

![image](https://user-images.githubusercontent.com/137158/33510482-5f7bdc4e-d6c1-11e7-9034-3064292bf66b.png)

I also noticed that my app wasn't requesting email permission:

<img width="302" alt="screen shot 2017-12-01 at 5 30 49 pm" src="https://user-images.githubusercontent.com/137158/33510494-7e5836da-d6c1-11e7-9e20-435c3d3ee024.png">

Yet the Graph API Explorer login mentions it explicitly:

![image](https://user-images.githubusercontent.com/137158/33510500-936078da-d6c1-11e7-81ba-8dbcc86e87d2.png)

Hope this helps somebody.

**Comment by mitianre** [comment]
The scope not being included as a parameter in the request URL is the problem, but I have no idea how to fix it.

I have something like this:
`https://www.facebook.com/v2.10/dialog/oauth?response_type=code&redirect_uri=REDIRECT_URI&client_id=CLIENT_ID`

instead of:
`https://www.facebook.com/v2.10/dialog/oauth?response_type=code&redirect_uri=REDIRECT_URI&scope=SCOPE_ITEMS&client_id=CLIENT_ID`

@statico, I've tried the same thing as you, but it's not working.

**Comment by mitianre** [comment]
Actually, I'm an idiot. That's what the problem was.

I didn't investigate the problem thoroughly and I've looked through the code only once in a while, when I had the time. Now I have seen that I was using a service over passport, that was calling the passport.authenticate method without passing options parameter, so of course the scope was empty.

Sorry for making such a "long" issue, but I hope that, somehow, the previous replies were helpful for others.

Idiot out. :))

**Comment by iudragon** [comment]
> I had a similar issue, with the new users which were unable to get the email field. Make sure when you get the permission request dialog the email is added there, otherwise you need to revoke the app permission and do it again.
> 
> Here are my settings:
> 
> ```
> scope:
> - 'email'
> - 'public_profile'
> profile_fields:
> - 'id'
> - 'email'
> - 'first_name'
> - 'last_name'
> - 'middle_name'
> - 'gender'
> - 'link'
> ```

Where exactly to add this fields in settings.py and in which section?

**Comment by JeffreyDavidsz** [comment]
@statico your solution fixed it for me. to make it a bit simpler;
This was the only way the email address was requested
`router.get('/auth/facebook', 
	passport.authenticate('facebook', {
		scope: ['email', 'public_profile']
	}
))`
and
`router.get('/auth/facebook/callback',
	passport.authenticate('facebook', {
		failureRedirect: '/error',
		successRedirect: '/'
	}))`
